server/aws-lsp-codewhisperer/src/language-server/amazonQServiceManager/AmazonQTokenServiceManager.ts

```diff
@@ -0,0 +1,403 @@
+        }
+
+        this.logServiceState('UpdateProfile is requested')
+
+        // Test if connection type changed
+        this.handleSsoConnectionChange()
+
+        if (this.connectionType !== 'identityCenter') {
+            this.logServiceState('Q Profile can not be set')
+            throw new AmazonQServiceNoProfileSupportError(
+                `Connection type ${this.connectionType} does not support profiles feature.`
+            )
+        }
+
+        if (this.state === 'INITIALIZED' && this.activeIdcProfile) {
+            // Change status to pending to prevent API calls until profile is updated.
+            // Because `listAvailableProfiles` below can take few seconds to complete,
+            // there is possibility that client could send requests while profile is changing.
+            this.state = 'PENDING_Q_PROFILE_UPDATE'
+        }
+
+        const profiles = await listAvailableProfiles(this.features)
+
+        const newProfile = profiles.find(el => el.arn === newProfileArn)
+
+        if (!newProfile) {
+            // TODO: do more validation if necessary, for now, checking only existence of the profile
+            // TODO: do we need to reset service here if requested profile does not exist anymore?
+            this.logging?.log(`Amazon Q Profile ${newProfileArn} is not valid`)
+            this.resetCodewhispererService()
+            this.state = 'PENDING_Q_PROFILE'
+
+            throw new AmazonQServiceInvalidProfileError('Requested Amazon Q Profile does not exist')
+        }
+
+        if (!this.activeIdcProfile) {
+            this.initializeCodewhispererService('identityCenter', newProfile.region)
+            this.activeIdcProfile = newProfile
+
+            return
+        }
+
+        // Profile didn't change
+        if (this.activeIdcProfile && this.activeIdcProfile.arn === newProfile.arn) {
+            // Update cached profile fields, keep existing client
+            this.logging?.log('Amazon Q Profile Change: profile selection did not change')
+            this.activeIdcProfile = newProfile
+            this.state = 'INITIALIZED'
+
+            return
+        }
+
+        // At this point new valid profile is selected.
+
+        const oldRegion = this.activeIdcProfile.region
+        const newRegion = newProfile.region
+        if (oldRegion === newRegion) {
+            this.logging?.log(`Amazon Q server: new profile is in the same region as old one, keeping exising service.`)
+            this.activeIdcProfile = newProfile
+            this.state = 'INITIALIZED'
+
+            return
+        }
+
+        this.logging?.log(`Amazon Q: switching service client region from ${oldRegion} to ${newRegion}`)
+
+        // Selected new profile is in different region. Re-initialize service and terminate inflight requests
+        this.resetCodewhispererService()
+        this.initializeCodewhispererService('identityCenter', newProfile.region)
+
+        this.activeIdcProfile = newProfile
```

Comment:
nit: can we set `this.state = 'INITIALIZED'` here not inside of `initializeCodewhispererService` method. It's more readable.